Track completion of non visible tasks as well

louwie17 opened the pull request:
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

Make sure we also track completion of non visible tasks and fix parent id in WooCommercePayments task

Closes #32140 

### How to test the changes in this Pull Request:

1. Run `pnpm nx build woocommerce` which will generate a zip
2. Load a Jurassic site with the zip, the [WooCommerce Admin Test Helper](https://github.com/woocommerce/woocommerce-admin-test-helper), and [WooCommerce Payments Dev Tools](https://github.com/Automattic/woocommerce-payments-dev-tools) (you probably want to disable the **Proxy WPCOM requests** setting in the WC Pay Dev Tools
3. Finish the onboarding wizard, installing WC Pay as part of the free business features
4. Enable the `progression_headercard` experiment under **WCA Test Helper > Experiments** enable the header card and the 2 col one to `treatment`
5. Go to **WooCommerce > Home** and finish the WooCommerce Payments task
6. Once connected The experiment task list won't show the Payments task anymore, but there should be a `wcadmin_tasklist_task_completed` generated to view this go to **WooCommerce > Status > Logs** and look for the `tracks-__current_date.log` log.

### Other information:

* [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully run tests with your changes locally?

<!-- Mark completed items with an [x] -->

### Changelog entry

> Track completion of non visible tasks within the task list.

### FOR PR REVIEWER ONLY:

* [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

louwie17 committed:
Make sure we also track completion of non visible tasks and fix parent id in WooCommercePayments task

louwie17 commented on plugins/woocommerce/src/Admin/Features/OnboardingTasks/Tasks/AdditionalPayments.php:
```diff
@@ -73,8 +73,11 @@ public function can_view() {
 
-		if ( ! $woocommerce_payments->is_requested() || ( $woocommerce_payments->is_supported() && ! $woocommerce_payments->is_connected() ) ) {
-			// Hide task if WC Pay is installed via OBW, in supported country, but not connected.
+		if ( ! $woocommerce_payments->is_requested() || ! $woocommerce_payments->is_supported() || ! $woocommerce_payments->is_connected() ) {
```
Matches condition of latest `3.3.1` patch release -> https://github.com/woocommerce/woocommerce-admin/pull/8506

louwie17 commented on plugins/woocommerce/src/Admin/Features/OnboardingTasks/Tasks/AdditionalPayments.php:
```diff
@@ -73,8 +73,11 @@ public function can_view() {
+			// Hide task if WC Pay is not installed via OBW, or is not connected, or the store is located in a country that is not supported by WC Pay.
+			return false;
+		}
+		if ( $this->get_parent_id() === 'extended_two_column' && WooCommercePayments::is_connected() ) {
```
Specific condition for the `extended_two_column` experimental task list

joshuatf replied:
Specific knowledge around the parent task list `extended_two_column` feels a tiny bit like a code smell, but I'm honestly not quite sure where else this would live at the moment.

louwie17 replied:
Yeah I would agree, we would almost just want a new Payments task class just for this task list with the updated logic, but that feels slightly like overkill.

louwie17 commented on plugins/woocommerce/src/Admin/Features/OnboardingTasks/Tasks/WooCommercePayments.php:
```diff
@@ -19,15 +19,6 @@ public function get_id() {
-	 */
-	public function get_parent_id() {
-		return 'setup';
-	}
```
Remove this as the `Task` class contains this method.

joshuatf replied:
Nice 👍 

Just a heads up that this override also exists in some old tests as well - https://github.com/woocommerce/woocommerce/blob/90a545b013c90580af2ce619d0b69f0d42faf29a/plugins/woocommerce/tests/legacy/unit-tests/woocommerce-admin/features/onboarding-tasks/test-task.php#L104-L106

louwie17 committed:
Remove test condition

joshuatf commented on plugins/woocommerce/src/Admin/Features/OnboardingTasks/TaskList.php:
```diff
@@ -350,6 +350,13 @@ public static function possibly_remove_reminder_bar() {
 	public function get_json() {
 		$this->possibly_track_completion();
+		// Track completion of non viewable tasks.
+		foreach ( $this->tasks as $task ) {
```
Nicely done!

What do you think about calling `get_json()` on all tasks below when we get the tasks?  We could then keep logic around tracking individual task completion to a minimum in the list and defer to the `Task` for tracking logic.

Then we can either filter them by viewable here or filter by viewable on the client if that's not too much work.  I think the latter is the way we should be handling this FWIW, but if this requires heavy changes on the frontend then we don't need to open that can of worms.

louwie17 replied:
I think that would be fine, my only caveat to that would be if the `get_json` triggers other expensive logic within tasks that are not viewable, meaning it would be kinda unnecessary.
But maybe that is something the actual task should handle anyway.

joshuatf replied:
Totally.  I share the same concerns, though my thinking is that `can_view` and `is_complete` are likely the expensive queries which we now need to run in both scenarios.

Using `get_json` was kind of a workaround to trigger tracks anyway and long-term we should probably refactor this slightly to possibly track completion in the constructor now that we're doing this for all tasks.

joshuatf reviewed (commented):
Code looks good and is testing well!  Tracks for non-viewable tasks also show in the new experiment.

Left a comment about moving the task track logic into the task, but otherwise this LGTM.

louwie17 committed:
Address PR feedback and fix PHP unit tests

louwie17 committed:
Add changelog

louwie17 commented:
Thanks for your review @joshuatf, I addressed your PR feedback in this commit -> https://github.com/woocommerce/woocommerce/pull/32493/commits/828e34a9745e5f84e195f76ed08e099048de4130
and responded to some of your comments.

louwie17 committed:
Update comment

louwie17 committed:
Move require_once below use statements

joshuatf reviewed (commented):
Thanks for the changes, @louwie17!  Everything looks good, though I think we'll also need to remove these lines to allow this to work:

https://github.com/woocommerce/woocommerce/blob/6336cb39f74d5b03ce14b2db34968aa9d3faf68d/plugins/woocommerce/src/Admin/Features/OnboardingTasks/Task.php#L361-L363

louwie17 committed:
Remove can_view condition on tracking completion

louwie17 commented:
> Thanks for the changes, @louwie17! Everything looks good, though I think we'll also need to remove these lines to allow this to work:
> 
> https://github.com/woocommerce/woocommerce/blob/6336cb39f74d5b03ce14b2db34968aa9d3faf68d/plugins/woocommerce/src/Admin/Features/OnboardingTasks/Task.php#L361-L363

Good catch @joshuatf, not sure how I missed that previously! I removed this in this commit -> https://github.com/woocommerce/woocommerce/pull/32493/commits/473d63d043a2a925b90a142454c4dd3e20fd9dd4

joshuatf reviewed (approved):
Great work, @louwie17!  LGTM 🎉
